### PR TITLE
Ccxt up

### PR DESCRIPTION
**Pull request by GuillaumeDSM**

### CODE REVIEW

File: octobot_trading/exchanges/connectors/ccxt/ccxt_clients_cache.py [hunk 1]
```diff
@@ -13,15 +13,21 @@
...
 
-from tests_additional.real_exchanges.test_huobi import TestHuobiRealExchangeTester
-# required to catch async loop context exceptions
-from tests import event_loop
+import octobot_commons.constants as commons_constants
 
-# All test coroutines will be treated as marked.
-pytestmark = pytest.mark.asyncio
 
+_MARKETS_BY_EXCHANGE = cachetools.TTLCache(maxsize=50, ttl=commons_constants.DAYS_TO_SECONDS*7)
```
Comment (Herklos) [inline octobot_trading/exchanges/connectors/ccxt/ccxt_clients_cache.py:21]: 👍

File: octobot_trading/exchanges/connectors/ccxt/ccxt_connector.py [hunk 2]
```diff
@@ -770,8 +787,8 @@ def get_max_handled_pair_with_time_frame(self) -> int:
         # 15 pairs, each on 3 time frames
         return 45
 
-    def supports_markets_as_raw_info(self) -> bool:
-        return ccxt_client_util.supports_markets_as_raw_info(self.client)
+    # def supports_markets_as_raw_info(self) -> bool:
+    #     return ccxt_client_util.supports_markets_as_raw_info(self.client)
```
Comment (Herklos) [inline octobot_trading/exchanges/connectors/ccxt/ccxt_connector.py:791]: Should we keep this code?

File: octobot_trading/exchanges/util/symbol_details.py [hunk 3]
```diff
@@ -21,8 +21,8 @@
 
 @dataclasses.dataclass
 class CCXTDetails(octobot_commons.dataclasses.FlexibleDataclass, octobot_commons.dataclasses.UpdatableDataclass):
-    info: dict = dataclasses.field(default_factory=dict)
-    parsed: dict = dataclasses.field(default_factory=dict)
+    info: dict = dataclasses.field(default_factory=dict)    # deprecated, will be removed soon
+    parsed: dict = dataclasses.field(default_factory=dict)  # deprecated, will be removed soon
```
Comment (Herklos) [inline octobot_trading/exchanges/util/symbol_details.py:25]: 👍

File: octobot_trading/personal_data/orders/decimal_order_adapter.py [hunk 4]
```diff
@@ -104,6 +107,29 @@ def decimal_adapt_order_quantity_because_price(limiting_value, max_value, price,
...
+) -> decimal.Decimal:
+    if side == enums.TradeOrderSide.BUY:
+        # only buy orders are affected
+        computed_fee = exchange_manager.exchange.get_trade_fee(
+            symbol, order_type, quantity, price, taker_or_maker.value
+        )
+        quote = commons_symbols.parse_symbol(symbol).quote
+        # if fee paid in quote, ensure enough remaining quote asset in available portfolio
+        if currency_fee := personal_data.get_fees_for_currency(computed_fee, quote):
+            fees_in_base = currency_fee / price if price else constants.ZERO
```
Comment (Herklos) [inline octobot_trading/personal_data/orders/decimal_order_adapter.py:123]: I don't think we should devide by zero when default value

File: octobot_trading/personal_data/orders/decimal_order_adapter.py [hunk 5]
```diff
@@ -104,6 +107,29 @@ def decimal_adapt_order_quantity_because_price(limiting_value, max_value, price,
...
+            base_quantity_from_quote = quote_available_funds / price
+            required_base_fees = fees_in_base * 2  # use 2x the fees to be sure exchanges won't refuse it
```
Comment (Herklos) [inline octobot_trading/personal_data/orders/decimal_order_adapter.py:125]: 👍

File: octobot_trading/util/test_tools/exchanges_test_tools.py [hunk 6]
```diff
@@ -110,15 +104,18 @@ async def _update_ohlcv(symbol):
...
         exchange_data.markets.append(details)
 
-    await exchange_manager.exchange.connector.load_symbol_markets(forced_markets=forced_markets)
+    await ensure_symbol_markets(exchange_manager, reload=reload_markets)
     await asyncio.gather(*(_update_ohlcv(symbol) for symbol in symbols))
     return exchange_data
 
 
+async def ensure_symbol_markets(exchange_manager, reload=False):
+    #todo call in backtesting
```
Comment (Herklos) [inline octobot_trading/util/test_tools/exchanges_test_tools.py:115]: Is it still a todo?

File: tests/personal_data/orders/test_decimal_order_adapter.py [hunk 7]
```diff
@@ -518,3 +520,82 @@ async def test_decimal_trunc_with_n_decimal_digits():
...
+    price = decimal.Decimal("30000")
+    taker_or_maker = enums.ExchangeConstantsMarketPropertyColumns.TAKER
+    quote_available_funds = decimal.Decimal("40000")
+
+    adapted = personal_data.decimal_adapt_order_quantity_because_fees(
+        exchange_manager, symbol, enums.TraderOrderType.BUY_LIMIT, origin_quantity,
+        price, taker_or_maker, enums.TradeOrderSide.SELL, quote_available_funds
+    )
+    # sell: no effect
+    assert adapted is origin_quantity
```
Comment (Herklos) [inline tests/personal_data/orders/test_decimal_order_adapter.py:553]: 👍

File: tests/personal_data/orders/test_decimal_order_adapter.py [hunk 8]
```diff
@@ -518,3 +520,82 @@ async def test_decimal_trunc_with_n_decimal_digits():
...
+    # buy and fees in USDT and not enough in quote_available_funds, quantity updated
+    assert adapted == origin_quantity - 2 * fees[enums.FeePropertyColumns.COST.value] / price
+
+    quote_available_funds = decimal.Decimal(100)
+    adapted = personal_data.decimal_adapt_order_quantity_because_fees(
+        exchange_manager, symbol, enums.TraderOrderType.BUY_LIMIT, origin_quantity,
+        price, taker_or_maker, enums.TradeOrderSide.BUY, quote_available_funds
+    )
+    # buy and fees in USDT and not enough in quote_available_funds, quantity greatly reduced
+    assert adapted == decimal.Decimal(100) / price - 2 * (fees[enums.FeePropertyColumns.COST.value] / price)
```
Comment (Herklos) [inline tests/personal_data/orders/test_decimal_order_adapter.py:593]: 👍

File: octobot_trading/personal_data/orders/decimal_order_adapter.py (hunk 4 above)
Comment (GuillaumeDSM) [reply octobot_trading/personal_data/orders/decimal_order_adapter.py:123]: indeed !

File: octobot_trading/util/test_tools/exchanges_test_tools.py (hunk 6 above)
Comment (GuillaumeDSM) [reply octobot_trading/util/test_tools/exchanges_test_tools.py:115]: nope, removed

File: octobot_trading/exchanges/connectors/ccxt/ccxt_connector.py (hunk 2 above)
Comment (GuillaumeDSM) [reply octobot_trading/exchanges/connectors/ccxt/ccxt_connector.py:791]: right, i'll remove it

File: tests_additional/real_exchanges/real_futures_exchange_tester.py [hunk 9]
```diff
@@ -22,6 +22,7 @@
 
 class RealFuturesExchangeTester(RealExchangeTester):
     EXCHANGE_TYPE = enums.ExchangeTypes.FUTURE.value
+    MARKET_STATUS_TYPE = "swap"
```
[the remark this answers is not in the document]
Comment (GuillaumeDSM) [reply tests_additional/real_exchanges/real_futures_exchange_tester.py:25]: it's the CCXT value. We don't use it in the bot, we use the "spot" one in a market filter. Here the idea is to make sure the "type" value of each market is set and has a real value

File: octobot_trading/exchanges/util/exchange_market_status_fixer.py [hunk 10]
```diff
@@ -277,11 +279,17 @@ def _get_price_precision(self):
 
     def _fix_market_status_precision_with_price(self):
         precision = self._get_price_precision()
-        self.market_status[Ecmsc.PRECISION.value] = {
-            Ecmsc.PRECISION_AMOUNT.value: precision,
-            Ecmsc.PRECISION_COST.value: precision,
-            Ecmsc.PRECISION_PRICE.value: precision,
-        }
```
Comment (Herklos) [inline octobot_trading/exchanges/util/exchange_market_status_fixer.py:284]: 😮

File: octobot_trading/exchanges/traders/trader.py [hunk 11]
```diff
@@ -286,6 +286,19 @@ async def _create_new_order(self, new_order, params: dict,
             await updated_order.state.wait_for_terminate(creation_timeout)
         return updated_order
 
+    def get_take_profit_order_type(self, base_order, order_type: enums.TraderOrderType) -> enums.TraderOrderType:
```
Comment (Herklos) [inline octobot_trading/exchanges/traders/trader.py:289]: 👍

**Review by Herklos — APPROVED** [review]
👍

**Review by Herklos — APPROVED** [review]

**Pull request merged**